### ISSUE TITLE
Not working in tf 2.0?

### ISSUE THREAD

**Issue by shreyaskamathkm92** [mhmoodlan/cyclic-learning-rate#1]
Hi,
I was trying to use cyclic lr in tf2.0. Below is my test code:

```
import tensorflow as tf
print(tf.__version__)
from debug import clr


mnist = tf.keras.datasets.mnist
(x_train, y_train), (_, _) = mnist.load_data()
x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
x_train = x_train / 255.0
y_train = y_train.astype('int32')

buffer_size = 5000
batch_size = 100
ds = tf.data.Dataset.from_tensor_slices((x_train, y_train)).shuffle(buffer_size)
ds = ds.batch(batch_size)

class My_Model(tf.keras.Model):
  def __init__(self):
    super(My_Model, self).__init__()
    self.flatten = tf.keras.layers.Flatten()
    self.fc1 = tf.keras.layers.Dense(10)

  def call(self, x, training=True):
    x = self.flatten(x)
    x = self.fc1(x)
    return x

def loss(logits, labels):
  return tf.reduce_mean(
      tf.nn.sparse_softmax_cross_entropy_with_logits(
          logits=logits, labels=labels))

step_counter = tf.Variable(0,  trainable = False, dtype=tf.int64)

learning_rate = clr.cyclic_learning_rate(step_counter,
                              learning_rate=0.01,
                              max_lr=0.1,
                              step_size=100,
                              gamma=0.99994,
                              mode='triangular',
                              name=None)
    

opt = tf.optimizers.Adam(learning_rate=learning_rate())

model = My_Model()

    
epochs = 2
for epoch in range(2):
  for batch, (images, labels) in enumerate(ds):
    with tf.GradientTape() as tape:
      logits = model(images, training=True)
      loss_value = loss(logits, labels)
      # print('loss', loss_value.numpy())
    grads = tape.gradient(loss_value, model.variables)
    opt.apply_gradients(zip(grads, model.trainable_variables))
    print('LR {}'.format(opt.learning_rate.numpy()))
    step_counter.assign_add(1)
    print('Step Counter  {}'.format(step_counter.numpy()))
```
For some reason, the learning rate is not being updated? Is there something wrong in the way it is being used?

Thank you,
Best,
SK

**Comment by mhmoodlan** [comment]
Hi SK,
I didn't test it -yet-, but yes, probably the current code is not compatible with tf2.0.
In addition, it is worth noting that github has been giving me security alerts about the tensorflow version that I am using in this code, therefore, use it with caution.
I will try to fix this as soon as I can.

**Comment by shreyaskamathkm92** [comment]
Hi,

Thank you so much for the reply. I think I got your code running on tf2.0. However, I did not test it yet. May be you can use this as a base and build on it. Below is the code:

```
import tensorflow as tf
from tensorflow.python.framework import ops
from tensorflow.python.ops import math_ops
from tensorflow.python.keras.optimizer_v2.learning_rate_schedule import LearningRateSchedule


class cyclic_learning_rate(LearningRateSchedule):
    def __init__(self,   learning_rate=0.01,
                         max_lr=0.1,
                         step_size=20.,
                         gamma=0.99994,
                         mode='triangular',
                         name=None):
        super(cyclic_learning_rate, self).__init__()
        self.learning_rate = learning_rate
        self.max_lr = max_lr
        self.step_size = step_size
        self.gamma = gamma
        self.mode = mode
        self.name = name
        
    def __call__(self, step):
        with ops.name_scope_v2(self.name or "CyclicLearningRate") as name:
            self.learning_rate = ops.convert_to_tensor(self.learning_rate, name="learning_rate")
            dtype = self.learning_rate.dtype
            global_step = math_ops.cast(step, dtype)
            self.step_size = math_ops.cast(self.step_size, dtype)
            
            double_step = math_ops.multiply(2., self.step_size)
            global_div_double_step = math_ops.divide(global_step, double_step)
            cycle = math_ops.floor(math_ops.add(1., global_div_double_step))
            # computing: x = abs( global_step / step_size – 2 * cycle + 1 )
            double_cycle = math_ops.multiply(2., cycle)
            global_div_step = math_ops.divide(global_step, self.step_size)
            tmp = math_ops.subtract(global_div_step, double_cycle)
            x = math_ops.abs(math_ops.add(1., tmp))
            # computing: clr = learning_rate + ( max_lr – learning_rate ) * max( 0, 1 - x )
            a1 = math_ops.maximum(0., math_ops.subtract(1., x))
            a2 = math_ops.subtract(self.max_lr, self.learning_rate)
            clr = math_ops.multiply(a1, a2)
            if self.mode == 'triangular2':
                clr = math_ops.divide(clr, math_ops.cast(math_ops.pow(2, math_ops.cast(cycle-1, tf.int32)), tf.float32))
            if self.mode == 'exp_range':
                clr = math_ops.multiply(math_ops.pow(self.gamma, global_step), clr)
            return math_ops.add(clr, self.learning_rate, name=name)

    def get_config(self):
        return {
            "initial_learning_rate": self.learning_rate,
            "max_learning_rate": self.max_lr,
            "step_size": self.step_size,
            "gamma": self.gamma,
            "mode": self.mode,
            "name": self.name
        }

```
